### ISSUE TITLE
[question] Reproducible build with build_type=Debug

### ISSUE THREAD

**Issue by Paultergeist** [conan-io/conan#16967]
### What is your question?

Documentation on [Revisions](https://docs.conan.io/2/tutorial/versioning/revisions.html#package-revisions) has the following paragraph:

> Best practices
> Having more than 1 package revision for any given recipe revision + `package_id` is a smell or a potential bad practice. It means that something was rebuilt when it was not necessary, wasting computing and storage resources. There are ways to avoid doing it, like `conan create . --build=missing:hello*` will only build that package binary if it doesn’t exist already (or running `conan graph info` can also return information of what needs to be built.)
> 

But because the build folder is different every time, Debug builds will always have different package revision even if nothing changes.

For example, if I create a new recipe with `conan new`:
```
conan new -d name=test_repro -d version=1.0 cmake_lib
```
and the build it twice with `build_type=Release`, package revision will be identical
```
% conan create . -s build_type=Release |& grep "Full package reference:"
test_repro/1.0: Full package reference: test_repro/1.0#4dd8581d5fb0f28ec7588863214486c4:8631cf963dbbb4d7a378a64a6fd1dc57558bc2fe#e61283aaa4c05b1b608229fcabe536b8
% conan create . -s build_type=Release |& grep "Full package reference:"
test_repro/1.0: Full package reference: test_repro/1.0#4dd8581d5fb0f28ec7588863214486c4:8631cf963dbbb4d7a378a64a6fd1dc57558bc2fe#e61283aaa4c05b1b608229fcabe536b8
```
But if I build with `build_type=Debug` twice in a row, I'll get two different revisions:
```
% conan create . -s build_type=Debug |& grep "Full package reference:"
test_repro/1.0: Full package reference: test_repro/1.0#4dd8581d5fb0f28ec7588863214486c4:451c431212a77cf562d97fe96ec83958609ba9e0#b853aa3f83f93ebf39cda69ffa5de8a1
% conan create . -s build_type=Debug |& grep "Full package reference:"
test_repro/1.0: Full package reference: test_repro/1.0#4dd8581d5fb0f28ec7588863214486c4:451c431212a77cf562d97fe96ec83958609ba9e0#e255e572311e32b0e5616e038df56493
```
My assumption that it's because build folder is different every time and path to sources gets embedded into binary as debug info.
```
% conan create . -s build_type=Debug |& grep "Build folder"
test_repro/1.0: Build folder /home/paul/.conan2/p/b/test_04e33249bdcc1/b/build/Debug
% conan create . -s build_type=Debug |& grep "Build folder"
test_repro/1.0: Build folder /home/paul/.conan2/p/b/test_94ae3a2fa0641/b/build/Debug
% conan create . -s build_type=Debug
<...skipped...>
test_repro/1.0: Package '451c431212a77cf562d97fe96ec83958609ba9e0' built
test_repro/1.0: Build folder /home/paul/.conan2/p/b/test_2137bcb049a84/b/build/Debug
test_repro/1.0: Generating the package
test_repro/1.0: Packaging in folder /home/paul/.conan2/p/b/test_2137bcb049a84/p
test_repro/1.0: Calling package()
test_repro/1.0: Running CMake.install()
test_repro/1.0: RUN: cmake --install "/home/paul/.conan2/p/b/test_2137bcb049a84/b/build/Debug" --prefix "/home/paul/.conan2/p/b/test_2137bcb049a84/p"
-- Install configuration: "Debug"
-- Installing: /home/paul/.conan2/p/b/test_2137bcb049a84/p/lib/libtest_repro.a
-- Installing: /home/paul/.conan2/p/b/test_2137bcb049a84/p/include/test_repro.h

test_repro/1.0: package(): Packaged 1 '.h' file: test_repro.h
test_repro/1.0: package(): Packaged 1 '.a' file: libtest_repro.a
test_repro/1.0: Created package revision 2f8c1fe52aab32626fbfbfa0710576c5
test_repro/1.0: Package '451c431212a77cf562d97fe96ec83958609ba9e0' created
test_repro/1.0: Full package reference: test_repro/1.0#4dd8581d5fb0f28ec7588863214486c4:451c431212a77cf562d97fe96ec83958609ba9e0#2f8c1fe52aab32626fbfbfa0710576c5
test_repro/1.0: Package folder /home/paul/.conan2/p/b/test_2137bcb049a84/p
% grep 2137bcb049a84 /home/paul/.conan2/p/b/test_2137bcb049a84/p/lib/libtest_repro.a
grep: /home/paul/.conan2/p/b/test_2137bcb049a84/p/lib/libtest_repro.a: binary file matches
```

So the question is - how to have reproducible builds with debug info? And why is build folder different every time?

The only way I see currently is to manually do 
```
conan install && conan build && conan export-pkg
```
but maybe there is better solution?

### Have you read the CONTRIBUTING guide?

- [X] I've read the CONTRIBUTING guide

**Comment by memsharded** [comment]
Hi @Paultergeist 

Thanks for your question.

The truth is that C++ are not deterministic in the general case, there are many other factors that can change the resulting binary beyond some source paths. So it could also easily affect Release builds as well. And they could be different package revisions even if the folder would be the same.

The real solution is not rebuilding again from source or recreating a package that already exists. This can be easily achieved with ``conan create ... --build=missing``.
In Conan 2, having more than 1 package revision for the same ``recipe-revision`` + ``package_id`` is considered a model or process error, and it should be avoided altogether in the first place (basically with the help of ``--build=missing``)

**Comment by memsharded** [comment]
Hi @Paultergeist 

Any further question or comment here?

**Comment by Paultergeist** [comment]
@memsharded 
Well, I got your point.
While I agree that source paths are not the only factor that affect the resulting binary, the other factors can be managed to achieve reproducible builds. There are practices of what to do and what not to do to get the same binary every time.
Currently, in my case, the randomized build folder is the only thing that prevents the build from being reproducible. I just wanted to check why the build folder path is different on every run, what's the rationale behind it and whether it can be fixed or not.

**Comment by memsharded** [comment]
>  I just wanted to check why the build folder path is different on every run, what's the rationale behind it and whether it can be fixed or not.

At the moment I don't see a way to be able to have always the same folder even before the build itself, because if the build is not guaranteed to be always reproducible, for all scenarios, re-using always the same build-folder will basically overwrite and destroy the previous package-revision, breaking traceability of package-revisions when builds are not fully deterministic (which unfortunately it is more common than the opposite)

Note there is also another opposing constraint, that is that the build artifacts should remain in the folder they were built for some cases, like debugging (see https://github.com/conan-io/conan/issues/13799) and packages that cannot be relocated after build (see https://github.com/conan-io/conan/issues/13237). Conan was relocating the build+package folder after the build when the package revision was available, very early in 2.0 initial iterations, but it was changed (see https://github.com/conan-io/conan/pull/13833) because of the above reasons.

So it basically seems impossible to satisfy both constraints simultaneously, and trying to change the current approach of creating a temporary folder with a ``uuid`` will break the other usages (I have just tried some quick attempts, until fully remembering about the above)

I am trying to think of some other alternative approach, like copying in the ``build()`` method all files from ``self.build_folder`` and ``self.source_folder`` elsewhere (fixed absolute location), building there and using that in the ``package()`` method. But this sounds quite fragile and ugly, I am not sure I'd recommend this, even if sounds possible.
